From 4d30a8b396e8112fc73447a7ea9483882c79b35e Mon Sep 17 00:00:00 2001
From: Jiri Benc <jbenc@redhat.com>
Date: Fri, 12 Sep 2014 14:30:15 -0400
Subject: [net] openvswitch: Avoid memory corruption in queue_userspace_packet()

Message-id: <4b83d999416213b3ef6ac150248a7cfb950e9649.1410531783.git.jbenc@redhat.com>
Patchwork-id: 94023
O-Subject: [RHEL7.1 net PATCH 42/50] openvswitch: Avoid memory corruption in queue_userspace_packet()
Bugzilla: 1110384
RH-Acked-by: Flavio Leitner <fbl@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110384
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7948291

commit f53e38317d581399eb67809d6b6b6c2c107db50c
Author: Andy Zhou <azhou@nicira.com>
Date:   Thu Jul 17 15:17:44 2014 -0700

    openvswitch: Avoid memory corruption in queue_userspace_packet()

    In queue_userspace_packet(), the ovs_nla_put_flow return value is
    not checked. This is fine as long as key_attr_size() returns the
    correct value. In case it does not, the current code may corrupt buffer
    memory. Add a run time assertion catch this case to avoid silent
    failure.

    Reported-by: Ben Pfaff <blp@nicira.com>
    Signed-off-by: Andy Zhou <azhou@nicira.com>
    Signed-off-by: Pravin B Shelar <pshelar@nicira.com>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/openvswitch/datapath.c b/net/openvswitch/datapath.c
index 014696b..a32d28a 100644
--- a/net/openvswitch/datapath.c
+++ b/net/openvswitch/datapath.c
@@ -466,7 +466,8 @@ static int queue_userspace_packet(struct datapath *dp, struct sk_buff *skb,
  upcall->dp_ifindex = dp_ifindex;
 
  nla = nla_nest_start(user_skb, OVS_PACKET_ATTR_KEY);
- ovs_nla_put_flow(upcall_info->key, upcall_info->key, user_skb);
+ err = ovs_nla_put_flow(upcall_info->key, upcall_info->key, user_skb);
+ BUG_ON(err);
  nla_nest_end(user_skb, nla);
 
  if (upcall_info->userdata)
-- 
1.7.1